### PR TITLE
GPS Sensor time

### PR DESCRIPTION
**Pull request by zyunlam**

### CODE REVIEW

File: MIDAS/src/hardware/GPSSensor.cpp [hunk 1]
```diff
@@ -17,17 +17,53 @@ ErrorCode GPSSensor::init() {
...
 }
 
+// This is needed because GPS doesn't provide unix time and just gives
+// dd mm yy
+uint16_t months[12] = {
+    0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334
+};
+
+inline bool is_leapyear(int year) {
+    return (!(year % 100 == 0) || (year % 400 == 0)) && year % 4 == 0;
```
Comment (redindelible) [inline MIDAS/src/hardware/GPSSensor.cpp:27]: `!(year % 100 == 0)` is sometimes called `year % 100 != 0`

File: MIDAS/src/hardware/GPSSensor.cpp [hunk 2]
```diff
@@ -17,17 +17,53 @@ ErrorCode GPSSensor::init() {
...
     uint16_t sat_count = gpgga_message.sats;
 
-    return GPS{lat, lon, alt, v, sat_count};
+    if (!is_leapyear(gprmc_message.date % 100)) {
+        is_leap = true;
+    }
+    if (!is_leap && is_leapyear(2000 + gprmc_message.date % 100)) {
+        is_leap = true;
+        for (int i = 2; i < 12; i++) {
+            months[i]++;
```
Comment (redindelible) [inline MIDAS/src/hardware/GPSSensor.cpp:58]: why are we modifying a global array

Comment (redindelible) [reply MIDAS/src/hardware/GPSSensor.cpp:58]: also this is fragile in the case that the rocket is turned on in december 2024 but is left running until is launched in february 2025

File: MIDAS/src/hardware/GPSSensor.cpp (hunk 1 above)
Comment (zyunlam) [reply MIDAS/src/hardware/GPSSensor.cpp:27]: I think !(year % 100) also works

File: MIDAS/src/hardware/GPSSensor.cpp [hunk 3]
```diff
@@ -19,12 +19,12 @@ ErrorCode GPSSensor::init() {
...
 // This is needed because GPS doesn't provide unix time and just gives
 // dd mm yy
-uint16_t months[12] = {
+const uint16_t months[12] = {
     0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334
 };
 
 inline bool is_leapyear(int year) {
-    return (!(year % 100 == 0) || (year % 400 == 0)) && year % 4 == 0;
+    return (!(year % 100) || (year % 400 == 0)) && year % 4 == 0;
```
Comment (redindelible) [inline MIDAS/src/hardware/GPSSensor.cpp:27]: i think it'd would have to be `(year % 100)`, not `!(year % 100)`. But please just use `(year % 100 != 0)`.

**Review by redindelible — CHANGES_REQUESTED** [review]

File: MIDAS/src/sensor_data.h [hunk 4]
```diff
@@ -104,11 +104,16 @@ struct Voltage {
...
+    int32_t latitude = 0;
+    int32_t longitude = 0;
     float altitude = 0;
     float speed = 0;
     uint16_t satellite_count = 0;
+    // Unix timestamp since 1970
+    // This isn't included in the telem packet because this is
+    // solely for the SD logger. We do not need to know what time it is
+    // when we are recieving telem packets.
+    uint32_t time = -1;
```
Comment (rbhog) [inline MIDAS/src/sensor_data.h:116]: nit: since it's a uint can we set it to smth within domain

Comment (rbhog) [reply MIDAS/src/sensor_data.h:116]: other option is to just not initialize it

**Review by aidancostello — CHANGES_REQUESTED** [review]
rebase off of main, we updated the telem packet to be a lot smaller.

**Review by aidancostello — APPROVED** [review]

**Review by redindelible — APPROVED** [review]

**Pull request merged**